src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,311 @@ private static string ToHtmlCase(string name)
+                {
+                    var op = Current;
+                    Next();
+
+                    if (op != '=' && At('='))
+                    {
+                        // Two length operator (ex: ^=). Move past the second piece
+                        Next();
+                    }
+
+                    return op;
+                }
+
+                return '\0';
+            }
+
+            private string ParseCSSValue(ErrorSink errorSink)
+            {
+                int valueStart, valueEnd;
+                if (At('\'') || At('"'))
+                {
+                    var quote = Current;
+
+                    // Move past the quote
+                    Next();
+
+                    valueStart = _index;
+                    var quoteEndIndex = _requiredAttributes.IndexOf(quote, _index);
+                    if (quoteEndIndex == -1)
+                    {
+                        errorSink.OnError(
+                            SourceLocation.Zero,
+                            Resources.FormatTagHelperDescriptorFactory_InvalidRequiredAttributeCharacterExpectedOther(
+                                _requiredAttributes[_requiredAttributes.Length - 1],
+                                quote),
+                            length: 0);
+                        return null;
+                    }
+                    valueEnd = quoteEndIndex;
+                    _index = quoteEndIndex + 1;
+                }
+                else
+                {
+                    valueStart = _index;
+                    var valueEndIndex = _requiredAttributes.IndexOfAny(InvalidCSSQuotelessValueCharacters, _index);
+                    valueEnd = valueEndIndex == -1 ? _requiredAttributes.Length : valueEndIndex;
+                    _index = valueEnd;
+                }
+
+                var value = _requiredAttributes.Substring(valueStart, valueEnd - valueStart);
+
+                return value;
+            }
+
+            private TagHelperRequiredAttributeDescriptor ParseCSSSelector(ErrorSink errorSink)
+            {
+                Debug.Assert(At('['));
+
+                // Move past '['.
+                Next();
+                PassOptionalWhitespace();
+
+                var attributeName = ParseCSSAttributeName(errorSink);
+
+                if (attributeName == null ||
+                    !ValidateName(attributeName, targetingAttributes: true, errorSink: errorSink))
+                {
+                    // Couldn't parse a valid attribute name.
+                    return null;
+                }
+
+                PassOptionalWhitespace();
+
+                if (!EnsureNotAtEnd(errorSink))
+                {
+                    return null;
+                }
+
+                var valueOperator = ParseCSSValueOperator(errorSink);
+
+                PassOptionalWhitespace();
+
+                if (!EnsureNotAtEnd(errorSink))
```

Comment:
Yeah, I dislike short helpers and `PassOptionalWhitespace()` would be longer and the boilerplate shorter if this bit were done in that method.